MobiusCore/Source/EffectHandlers/EffectHandler.swift

```diff
@@ -16,98 +16,53 @@
+/// type. This process is not necessarily synchronous and that any number of `Event`s could be outputted for a given `Effect`.
+///
+/// The `connect` function is called when an `EffectHandler` should start handling effects. `stopHandling` will be called when this connection is torn
+/// down. You must therefore dispose of any resources you are using when you receive this callback.
+///
+/// Note: Sending events from your `EffectHandler` after `stopHandling` returns __will__ cause a runtime error.
+public struct EffectHandler<Effect, Event> {
+    private let connectFn: (@escaping Consumer<Event>) -> _EffectHandlerConnection<Effect, Event>
+    fileprivate init(
+        connect: @escaping (@escaping Consumer<Event>) -> _EffectHandlerConnection<Effect, Event>
     ) {
-        disposeFn = disposable
-        self.handleEffect = { effect in
-            if let payload = canHandle(effect) {
-                return { dispatch in
-                    handle(payload, dispatch)
-                }
-            } else {
-                return nil
-            }
-        }
-    }
-
-    /// Connect to this `EffectHandler` by supplying it with an output for its events.
-    /// This will return a `Connection` which you can use to send effects to this `EffectHandler` and to tear down the connection.
-    ///
-    /// NOTE: Only one connection can be held to this `EffectHandler` at a time.
-    ///
-    /// - Parameter consumer: the output that this `EffectHandler` should send its events to.
-    public func connect(_ consumer: @escaping (Event) -> Void) -> Connection<Effect> {
-        return lock.synchronized {
-            guard self.consumer == nil else {
-                MobiusHooks.onError("An EffectHandler only supports one connection at a time.")
-                return BrokenConnection<Effect>.connection()
-            }
-            self.consumer = consumer
-
-            return Connection(
-                acceptClosure: self.accept,
-                disposeClosure: self.dispose
-            )
-        }
-    }
-
-    func canAccept(_ effect: Effect) -> Bool {
-        return handleEffect(effect) != nil
-    }
-
-    private func accept(_ effect: Effect) {
-        if let performEffect = handleEffect(effect) {
-            performEffect { [unowned self] event in
-                self.dispatch(event: event)
-            }
-        }
+        connectFn = connect
     }
 
-    private func dispatch(event: Event) {
-        return lock.synchronized {
-            guard let consumer = self.consumer else {
-                MobiusHooks.onError("Nothing is connected to this `EffectHandler`. Ensure your resources have been cleaned up in `stopHandling`")
-                return
-            }
-
-            consumer(event)
-        }
+    /// Start handling effects.
+    /// An `EffectHandlerConnection` is returned which can handle effects and can be torn down when it should stop.
```

Comment:
The class must be public since it is exposed by the API. The underscore (and use of `final`) is to signal that it should not be of interest to the user of the API.
I will clarify this in a comment.